localstack-core/localstack/config.py

```diff
@@ -584,40 +632,56 @@ class UniqueHostAndPortList(List[HostAndPort]):
     """
     Container type that ensures that ports added to the list are unique based
     on these rules:
-        - 0.0.0.0 "trumps" any other binding, i.e. adding 127.0.0.1:4566 to
-          [0.0.0.0:4566] is a no-op
-        - adding identical hosts and ports is a no-op
-        - adding `0.0.0.0:4566` to [`127.0.0.1:4566`] "upgrades" the binding to
-          create [`0.0.0.0:4566`]
+        - :: "trumps" any other binding on the same port, including both IPv6 and IPv4
+          addresses. All other bindings for this port are removed, since :: already
+          covers all interfaces. For example, adding 127.0.0.1:4566, [::1]:4566,
+          and [::]:4566 would result in only [::]:4566 being preserved.
+        - 0.0.0.0 "trumps" any other binding on IPv4 addresses only. IPv6 addresses
+          are not removed.
+        - Identical identical hosts and ports are de-duped
     """
 
-    def __init__(self, iterable=None):
-        super().__init__()
-        for item in iterable or []:
-            self.append(item)
+    def __init__(self, iterable: List[HostAndPort] | None = None):
```

Comment:
I made a new algorithm for doing the dedupe, where the entire list is reconciled each time the list is modified, instead of having two separate sets of rules depending on the order items are added.

This also handles some additional corner cases, such as if the list contains multiple different IPs on the same port, and then a single `::` or `0.0.0.0` is added with the same port (which would previously result in a duplicate entry).